dog.go

```diff
@@ -0,0 +1,200 @@
+
+// zombieSafeRadius: if a zombie is at least this far from the dog, it stops running
+const zombieSafeRadius float64 = 192
+
+// states of the dog
+// It would be great to map them to the frameTag.Name from JSON
+const (
+	dogWalking  int = 0
+	dogSniffing     = 1
+	dogSitting      = 2
+)
+
+// Dog is player's companion
+type Dog struct {
+	Object   *resolv.Object
+	Angle    float64
+	Speed    float64
+	Frame    int
+	State    int
+	Path     Path
+	NextPath int
+	Sprite   *SpriteSheet
+	InDanger bool
+}
+
+// Update updates the state of the dog
+func (d *Dog) Update(g *Game) {
+	isSafeAgain := true
+	resultantVector := Coord{
+		X: 0,
+		Y: 0,
+	}
+	for _, zombie := range g.Zombies {
+		zombieDistance, xDistance, yDistance := CalcObjectDistance(d.Object, zombie.Object)
+		if zombieDistance < zombieDangerRadius {
+			// If zombies are too close then the dog is in danger and will run away
+			resultantVector.X += xDistance
+			resultantVector.Y += yDistance
+			isSafeAgain = false
+			d.InDanger = true
+			d.State = dogWalking
+		} else if d.InDanger && zombieDistance < zombieSafeRadius {
+			// If the dog is running away from zombies then it will be safe again when getting far enough from the zombies
+			isSafeAgain = false
+		}
+	}
+
+	// If the dog is not in danger anymore then it turn towards the next path point
+	if d.InDanger && isSafeAgain {
+		d.TurnTowardsPathPoint()
+	}
+	d.InDanger = !isSafeAgain
+
+	if (!d.InDanger) {
+		playerDistance, _, _ := CalcObjectDistance(d.Object, g.Player.Object)
+		if playerDistance < waitingRadius {
+			// If the dog is not in danger and it is close to the player then it sniffs towards next path point
+			d.State = dogSniffing
+			d.FollowPath()
+		} else {
+		// If the player is not close enough then the dog sits down
+		d.State = dogSitting
+		}
+	} else {
+		// If the dog is in danger then it runs away from the zombies
+		d.TurnTowardsCoordinate(resultantVector)
+		d.Run()
+	}
+
+	d.animate(g)
+	d.Object.Update()
+}
+
+func (d *Dog) animate(g *Game) {
+	// Update only in every 5th cycle
+	if g.Tick%5 != 0 {
+		return
+	}
+
+	ft := d.Sprite.Meta.FrameTags[d.State]
+
+	if ft.From == ft.To {
+		d.Frame = ft.From
+	} else {
+		// Contiuously increase the Frame counter between From and To
+		d.Frame = (d.Frame-ft.From+1)%(ft.To-ft.From+1) + ft.From
+	}
+}
+
+// TurnTowardsCoordinate turns the dog to the direction of the point
+func (d *Dog) TurnTowardsCoordinate(coord Coord) {
+	adjacent := d.Object.X - coord.X
+	opposite := d.Object.Y - coord.Y
+	// math.Pi is needed only until the dog sprite is looking up
+	d.Angle = math.Atan2(opposite, adjacent)+math.Pi/2
+}
+
+// TurnTowardsPathPoint turns the dog to the direction of the next path point
+func (d *Dog) TurnTowardsPathPoint() {
+	adjacent := d.Object.X - float64(d.Path[d.NextPath].X)
+	opposite := d.Object.Y - float64(d.Path[d.NextPath].Y)
+	// math.Pi is needed only until the dog sprite is looking up
+	d.Angle = math.Atan2(opposite, adjacent)-math.Pi
```

Comment:
Why is this one -math.Pi and the other one is +math.Pi/2? 🤔

Reply:
I am updating the sprites and then will check and correct all the trigonometry.
I think it is because in one case the vector pointing towards the dog, while in the other case it is pointing away from the dog

Reply:
But it works :)